### PR TITLE
Polish Signal based operators

### PR DESCRIPTION
**Pull request by smaldini**
- Turn Signal into interface
- Remove MutableNextSignal
- Precise FluxPeekStateful to FluxDoOnEach

### CODE REVIEW

File: reactor-core/src/main/java/reactor/core/publisher/ImmutableSignal.java [hunk 1]
```diff
@@ -69,4 +70,72 @@ public SignalType getType() {
 		return type;
 	}
 
+	//the base class defines equals and hashcode as final in order to allow
+
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/ImmutableSignal.java]: equals is not final and this is not a base class anymore, so remove broken comment ;)

File: reactor-core/src/main/java/reactor/core/publisher/ImmutableSignal.java [hunk 2]
```diff
@@ -69,4 +70,72 @@ public SignalType getType() {
...
+		}
+		if (isOnError()) {
+			return Objects.equals(this.getThrowable(), signal.getThrowable());
+		}
+		if (isOnNext()) {
+			return Objects.equals(this.get(), signal.get());
+		}
+		return false;
+	}
+	//to implement additional state.
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/ImmutableSignal.java]: that was part of the comment of equals, remove it too

**Comment by smaldini** [comment]
The rationale is simply shrinking surface API when possible due to its specificity and simplicity to reproduce. Also the operator was only used by doOnEach and if we happen to implement a doOnNext with state we'll revisit. In fact this operator should also be used by ParallelFlux too and eventually be fuseable.

**Comment by smaldini** [comment]
Also DoOnEachSubscriber implement Signal directly to avoid allocations.

**Pull request merged**